### PR TITLE
Add babelfishpg_tsql.dump_restore GUC to support pg_dump/restore

### PR DESCRIPTION
**Pull request by 2jungkook**
pg_dump generates DDLs in postgres dialect and pg_restore executes those DDLs also in postgres dialect. However, in some cases, e.g., a view having collatable constants, we need a special handling to preserve T-SQL behavior. This commit introduces a new GUC babelfishpg_tsql.dump_restore that can be set during pg_dump and then executed during pg_restore.

Signed-off-by: Jungkook Lee <jungkook@amazon.com>

### CODE REVIEW

**Review by rishabhtanwar29 — APPROVED** [review]
We should add a good commit message while merging this PR.

File: contrib/babelfishpg_tsql/src/guc.c [hunk 1]
```diff
@@ -69,6 +69,9 @@ static const struct config_enum_entry explain_format_options[] = {
 
 extern bool Transform_null_equals;
 
+/* Dump and Restore */
+bool babelfish_restore_mode = false;
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/guc.c:73]: if it is boolean, we should not use mode (implies enum/string with different values). Also, restore is too generic and might create confusion. May be babelfish_dump_restore/upgrade_restore?

File: contrib/babelfishpg_tsql/src/guc.c [hunk 2]
```diff
@@ -979,6 +982,16 @@ define_custom_variables(void)
...
 				 GUC_NOT_IN_SAMPLE,
 				 NULL, NULL, NULL);
+
+	/* Dump and Restore */
+	DefineCustomBoolVariable("babelfishpg_tsql.restore_mode",
+				 gettext_noop("enable babelfish restore mode"),
+				 NULL,
+				 &babelfish_restore_mode,
+				 false,
+				 PGC_USERSET,
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/guc.c:992]: I think we should not allow user to set/reset this GUC, can we use a permission which is most restrictive in that sense

File: contrib/babelfishpg_tsql/src/guc.c (hunk 1 above)
Comment (2jungkook) [reply contrib/babelfishpg_tsql/src/guc.c:73]: Okay, let me change the name to `babelfish_dump_restore`.

File: contrib/babelfishpg_tsql/src/guc.c (hunk 2 above)
Comment (2jungkook) [reply contrib/babelfishpg_tsql/src/guc.c:992]: Any user can do dump/restore, can't they?

Comment (surendravishnoi) [reply contrib/babelfishpg_tsql/src/guc.c:992]: that is what I meant, a dump or restore can set this option but not just any user. Can we enforce that option is only changed by dump/restore code (which we control)

Comment (2jungkook) [reply contrib/babelfishpg_tsql/src/guc.c:992]: Actually, pg_dump just generates SQL texts, e.g., `SET babelfish_dump_restore TO ON`. However, a user can do restore by either using pg_restore or psql, i.e.,
- If `--format=custom` is given during pg_dump, then the user should use pg_restore
- If `--format=plain` is given during pg_dump, then the user should use `psql -f dumped.sql`

So I don't think there is an easy way to enforce this option to the specific context.

Comment (surendravishnoi) [reply contrib/babelfishpg_tsql/src/guc.c:992]: ok, atleast we should hide it from show

**Review by surendravishnoi — APPROVED** [review]

**Pull request merged**